.github/workflows/integration.yml

```diff
@@ -0,0 +1,55 @@
+name: hashicorp/waypoint-hzn/integration
+on:
+  push:
+    # TODO: do we want to push a new image everytime we push to main (including this new PR)?
+    branches:
+      # - main
+      - convert-hashicorp-waypoint-hzn-to-actions-20230412-181424
+jobs:
+  dev-build:
+    runs-on: ubuntu-latest
+    outputs:
+      go-version: ${{ steps.go-version.outputs.go-version }}
+    steps:
+      - uses: actions/checkout@83b7061638ee4956cf7545a6f7efe594e5ad0247 # v3.5.1
+      - uses: actions/setup-go@6edd4406fa81c3da01a34fa6f6343087c207a568 # https://github.com/actions/setup-go/releases/tag/v3.5.0
+        with:
+          go-version-file: go.mod
+      - id: go-version
+        run: echo "go-version=$(cat ./.go-version)" >> "$GITHUB_OUTPUT"
+      - uses: hashicorp/actions-go-build@v0.1.7
+        with:
+          go_version: ${{ steps.go-version.outputs.go-version }}
+          os: linux
+          arch: amd64
+          reproducible: nope
+          # TODO: inquire about versioning
+          instructions: |-
+            mkdir -p dist
+            go build -o dist ./cmd/waypoint-hzn
+
+  image-release:
+    runs-on: ubuntu-latest
+    needs:
+      - dev-build
+    steps:
+      - uses: actions/checkout@83b7061638ee4956cf7545a6f7efe594e5ad0247 # v3.5.1
+      - uses: actions/download-artifact@9bc31d5ccc31df68ecc42ccf4149144866c47d8a # v3.0.2
+        with:
+          path: waypoint-hzn
+      - name: Log in to the Container registry
+        uses: docker/login-action@f054a8b539a109f9f41c372932f1ae047eff08c9
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      # TODO: uncomment this block to enable pushes to ghcr
```

Comment:
Note that if we enable a new image to be released on every push to main, we this Circle CI -> GHA migration will initiate a new release.